Non-deterministic test failure in test_data_at_range_limit

I noticed the following failure when working on #10850 - this seems to happen on manylinux1 64-bit Python. I'll try and provide docker commands to reproduce later.

```
=================================== FAILURES ===================================
______________ test_data_at_range_limit[False-fast_reader1-True] _______________

parallel = False, fast_reader = {'parallel': False, 'use_fast_converter': False}
guess = True

    @pytest.mark.parametrize("guess", [True, False])
    # fast_reader configurations: False| 'use_fast_converter'=False|True
    @pytest.mark.parametrize('fast_reader', [False, dict(use_fast_converter=False),
                                             dict(use_fast_converter=True)])
    @pytest.mark.parametrize("parallel", [False, True])
    def test_data_at_range_limit(parallel, fast_reader, guess):
        """
        Test parsing of fixed-format float64 numbers near range limits
        (|~4.94e-324 to 1.7977e+308|) - within limit for full precision
        (|~2.5e-307| for strtod C parser, factor 10 better for fast_converter)
        exact numbers shall be returned, beyond that an Overflow warning raised.
        Input of exactly 0.0 must not raise an OverflowError.
        """
        # Python reader and strtod() are expected to return precise results
        rtol = 1.e-30
    
        # Update fast_reader dict; adapt relative precision for fast_converter
        if fast_reader:
            fast_reader['parallel'] = parallel
            if fast_reader.get('use_fast_converter'):
                rtol = 1.e-15
            elif np.iinfo(np.int_).dtype == np.dtype(np.int32):
                # On 32bit the standard C parser (strtod) returns strings for these
                pytest.xfail("C parser cannot handle float64 on 32bit systems")
    
        if parallel:
            if not fast_reader:
                pytest.skip("Multiprocessing only available in fast reader")
            elif TRAVIS:
                pytest.xfail("Multiprocessing can sometimes fail on Travis CI")
    
        # Test very long fixed-format strings (to strtod range limit w/o Overflow)
        for D in 99, 202, 305:
            t = ascii.read(StringIO(99 * '0' + '.' + D * '0' + '1'), format='no_header',
                           guess=guess, fast_reader=fast_reader)
            assert_almost_equal(t['col1'][0], 10.**-(D + 1), rtol=rtol, atol=1.e-324)
        for D in 99, 202, 308:
            t = ascii.read(StringIO('1' + D * '0' + '.0'), format='no_header',
                           guess=guess, fast_reader=fast_reader)
            assert_almost_equal(t['col1'][0], 10.**D, rtol=rtol, atol=1.e-324)
    
        # 0.0 is always exact (no Overflow warning)!
        for s in '0.0', '0.0e+0', 399 * '0' + '.' + 365 * '0':
            with pytest.warns(None) as w:
                t = ascii.read(StringIO(s), format='no_header',
                               guess=guess, fast_reader=fast_reader)
            assert t['col1'][0] == 0.0
            assert len(w) == 0
    
        # Test OverflowError at precision limit with laxer rtol
        if parallel:
            pytest.skip("Catching warnings broken in parallel mode")
        elif not fast_reader:
            pytest.skip("Python/numpy reader does not raise on Overflow")
        with pytest.warns(AstropyWarning, match=r'OverflowError converting to '
                          r'FloatType in column col1, possibly resulting in '
                          r'degraded precision'):
            t = ascii.read(StringIO('0.' + 314 * '0' + '1'), format='no_header',
>                          guess=guess, fast_reader=fast_reader)
E           Failed: DID NOT WARN. No warnings of type (<class 'astropy.utils.exceptions.AstropyWarning'>,) was emitted. The list of emitted warnings is: [].

/tmp/tmp.siYQiS1424/venv/lib/python3.6/site-packages/astropy/io/ascii/tests/test_c_reader.py:1181: Failed
______________ test_data_at_range_limit[False-fast_reader1-False] ______________

parallel = False, fast_reader = {'parallel': False, 'use_fast_converter': False}
guess = False

    @pytest.mark.parametrize("guess", [True, False])
    # fast_reader configurations: False| 'use_fast_converter'=False|True
    @pytest.mark.parametrize('fast_reader', [False, dict(use_fast_converter=False),
                                             dict(use_fast_converter=True)])
    @pytest.mark.parametrize("parallel", [False, True])
    def test_data_at_range_limit(parallel, fast_reader, guess):
        """
        Test parsing of fixed-format float64 numbers near range limits
        (|~4.94e-324 to 1.7977e+308|) - within limit for full precision
        (|~2.5e-307| for strtod C parser, factor 10 better for fast_converter)
        exact numbers shall be returned, beyond that an Overflow warning raised.
cibuildwheel-26d88921-1f51-4163-9e4e-f61fe6b3c83b
        Input of exactly 0.0 must not raise an OverflowError.
        """
        # Python reader and strtod() are expected to return precise results
        rtol = 1.e-30
    
        # Update fast_reader dict; adapt relative precision for fast_converter
        if fast_reader:
            fast_reader['parallel'] = parallel
            if fast_reader.get('use_fast_converter'):
                rtol = 1.e-15
            elif np.iinfo(np.int_).dtype == np.dtype(np.int32):
                # On 32bit the standard C parser (strtod) returns strings for these
                pytest.xfail("C parser cannot handle float64 on 32bit systems")
    
        if parallel:
            if not fast_reader:
                pytest.skip("Multiprocessing only available in fast reader")
            elif TRAVIS:
                pytest.xfail("Multiprocessing can sometimes fail on Travis CI")
    
        # Test very long fixed-format strings (to strtod range limit w/o Overflow)
        for D in 99, 202, 305:
            t = ascii.read(StringIO(99 * '0' + '.' + D * '0' + '1'), format='no_header',
                           guess=guess, fast_reader=fast_reader)
            assert_almost_equal(t['col1'][0], 10.**-(D + 1), rtol=rtol, atol=1.e-324)
        for D in 99, 202, 308:
            t = ascii.read(StringIO('1' + D * '0' + '.0'), format='no_header',
                           guess=guess, fast_reader=fast_reader)
            assert_almost_equal(t['col1'][0], 10.**D, rtol=rtol, atol=1.e-324)
    
        # 0.0 is always exact (no Overflow warning)!
        for s in '0.0', '0.0e+0', 399 * '0' + '.' + 365 * '0':
            with pytest.warns(None) as w:
                t = ascii.read(StringIO(s), format='no_header',
                               guess=guess, fast_reader=fast_reader)
            assert t['col1'][0] == 0.0
            assert len(w) == 0
    
        # Test OverflowError at precision limit with laxer rtol
        if parallel:
            pytest.skip("Catching warnings broken in parallel mode")
        elif not fast_reader:
            pytest.skip("Python/numpy reader does not raise on Overflow")
        with pytest.warns(AstropyWarning, match=r'OverflowError converting to '
                          r'FloatType in column col1, possibly resulting in '
                          r'degraded precision'):
            t = ascii.read(StringIO('0.' + 314 * '0' + '1'), format='no_header',
>                          guess=guess, fast_reader=fast_reader)
E           Failed: DID NOT WARN. No warnings of type (<class 'astropy.utils.exceptions.AstropyWarning'>,) was emitted. The list of emitted warnings is: [].

/tmp/tmp.siYQiS1424/venv/lib/python3.6/site-packages/astropy/io/ascii/tests/test_c_reader.py:1181: Failed
=========================== short test summary info ============================

--- a/astropy/io/ascii/tests/test_c_reader.py
+++ b/astropy/io/ascii/tests/test_c_reader.py
@@ -1174,11 +1174,15 @@ def test_data_at_range_limit(parallel, fast_reader, guess):
         pytest.skip("Catching warnings broken in parallel mode")
     elif not fast_reader:
         pytest.skip("Python/numpy reader does not raise on Overflow")
-    with pytest.warns(AstropyWarning, match=r'OverflowError converting to '
-                      r'FloatType in column col1, possibly resulting in '
-                      r'degraded precision'):
+    with pytest.warns(None) as warning_lines:
         t = ascii.read(StringIO('0.' + 314 * '0' + '1'), format='no_header',
                        guess=guess, fast_reader=fast_reader)
+
+    n_warns = len(warning_lines)
+    assert n_warns in (0, 1), f'Expected 0 or 1 warning, found {n_warns}'
+    if n_warns == 1:
+        assert 'OverflowError converting to FloatType in column col1, possibly resulting in degraded precision' in str(warning_lines[0].message)  # noqa
+
     assert_almost_equal(t['col1'][0], 1.e-315, rtol=1.e-10, atol=1.e-324)
 
 

